### PR TITLE
Separate tests for exceptions 

### PR DESCRIPTION
**Pull request by kamathhrishi**
This PR focuses on solving issue #108. The issue is to add tests to check if the right exceptions are thrown for incorrect behavior. 

- [x] Test if an exception is when the wrong approximation method is used for Tanh 
- [x]  Test if an exception is when the wrong approximation method is used for reciprocal
- [x]  Test if an exception is thrown when session is not passed to MPC tensor 
- [x]  Test if an exception is thrown when the session is not for passed for the torch.tensor(<value>).share()
- [x]   Test if an exception is thrown when not enough primitives for FSS
- [x]   Test if an exception is thrown when kernel mismatch occurs in convolutions 
- [x]    Test if an exception is thrown when ring size is mismatched 
- [x]   Test if an exception is thrown when a protocol not present is called the session 
- [x]  Test if an exception is thrown when the user performs division on floating-point numbers
- [x]  Test if an exception is thrown when division is performed on MPCTensor  
- [x] Test if an exception is thrown when an unsupported operation is performed 
- [x] Test if an exception is thrown when an unsupported FSS operation is performed (when generating primitives)

### CODE REVIEW

File: tests/sympc/approximations/reci_test.py [hunk 1]
```diff
@@ -22,5 +22,15 @@ def test_reciprocal(method, get_clients) -> None:
     x_reciprocal = reciprocal(x, method=method)
     assert torch.allclose(x_secret_reciprocal, x_reciprocal.reconstruct(), atol=1e-1)
 
+
+def test_exception(get_clients) -> None:
```
Comment (gmuraru) [inline tests/sympc/approximations/reci_test.py:26]: MAYBE: Could you also add the exception name to the test?
like: ```test_exception_value_error```?

File: tests/sympc/approximations/tanh_test.py [hunk 2]
```diff
@@ -23,3 +23,17 @@ def test_tanh(get_clients) -> None:
 
     with pytest.raises(ValueError):
         x_tanh = tanh(x, method="exp")
+
+
+def test_exception(get_clients) -> None:
```
Comment (gmuraru) [inline tests/sympc/approximations/tanh_test.py:28]: Also for this

File: tests/sympc/session/session_test.py [hunk 3]
```diff
@@ -77,3 +78,8 @@ def test_eq():
     assert session != other1
     # Test equal sessions:
     assert session == other2
+
+
+def test_protocol_exception():
+    with pytest.raises(ValueError):
+        session = Session(protocol="fs")
```
Comment (gmuraru) [inline tests/sympc/session/session_test.py:85]: :+1:

**Review by gmuraru — COMMENTED** [review]
It looks really great! You are doing a good job!

File: tests/sympc/approximations/reci_test.py (hunk 1 above)
Comment (kamathhrishi) [reply tests/sympc/approximations/reci_test.py:26]: Sure

File: tests/sympc/tensor/mpc_tensor_test.py [hunk 4]
```diff
@@ -178,6 +203,34 @@ def test_ops_mpc_public(get_clients, nr_clients, op_str) -> None:
     assert np.allclose(result, expected_result, atol=10e-4)
 
 
+def test_ops_divfloat_exception() -> None:
```
Comment (gmuraru) [inline tests/sympc/tensor/mpc_tensor_test.py:206]: MAYBE: There is a `get_clients` fixture in `conftest.py`. Can't you use that for clients?

Comment (kamathhrishi) [reply tests/sympc/tensor/mpc_tensor_test.py:206]: Oh, I can. I will make the change.

File: tests/sympc/protocol/protocol_test.py [hunk 5]
```diff
@@ -8,3 +13,15 @@ class TestProtocol(metaclass=Protocol):
...
+def test_exception_unsupported_fss_operation():
+    with pytest.raises(ValueError):
+        _generate_primitive("relu", 10)
+
+
+def test_exception_insufficient_fss_primitives():
+    store = {}
+    store["fss_eq"] = [[10]]
+    with pytest.raises(ValueError):
+        _get_primitive("fss_eq", store, 100)
```
Comment (gmuraru) [inline tests/sympc/protocol/protocol_test.py:27]: MAYBE: Could you use the `get_primitive` function (from the same module `sympc.protocol.fss.fss`).
The functions that start with `_` are meant to be private.

File: tests/sympc/protocol/protocol_test.py [hunk 6]
```diff
@@ -8,3 +13,15 @@ class TestProtocol(metaclass=Protocol):
 def test_register_protocol() -> None:
     assert "TestProtocol" in Protocol.registered_protocols
     assert Protocol.registered_protocols["TestProtocol"] == TestProtocol
+
+
+def test_exception_unsupported_fss_operation():
+    with pytest.raises(ValueError):
+        _generate_primitive("relu", 10)
```
Comment (gmuraru) [inline tests/sympc/protocol/protocol_test.py:20]: Same comment from bellow, only that `generate_primitive`

File: tests/sympc/tensor/mpc_tensor_test.py [hunk 7]
```diff
@@ -23,6 +23,29 @@ def test_mpc_tensor_exception(get_clients) -> None:
         x = MPCTensor(secret=torch.Tensor([1, -2]), session=session)
 
 
+def test_mpc_tensor_nosession_exception(get_clients) -> None:
+    alice_client, bob_client = get_clients(2)
```
Comment (gmuraru) [inline tests/sympc/tensor/mpc_tensor_test.py:27]: MAYBE: Remove this

File: tests/sympc/tensor/mpc_tensor_test.py [hunk 8]
```diff
@@ -23,6 +23,29 @@ def test_mpc_tensor_exception(get_clients) -> None:
...
+
+    with pytest.raises(ValueError):
+        MPCTensor(secret=42)
+
+    with pytest.raises(ValueError):
+        x = MPCTensor(secret=torch.Tensor([1, -2]))
+
+
+def test_mpc_share_nosession_exception(get_clients) -> None:
+    alice_client, bob_client = get_clients(2)
```
Comment (gmuraru) [inline tests/sympc/tensor/mpc_tensor_test.py:37]: MAYBE: Remove this

File: tests/sympc/tensor/mpc_tensor_test.py [hunk 9]
```diff
@@ -178,6 +202,27 @@ def test_ops_mpc_public(get_clients, nr_clients, op_str) -> None:
     assert np.allclose(result, expected_result, atol=10e-4)
 
 
+def test_ops_divfloat_exception(get_clients) -> None:
+    # Define the virtual machines that would be use in the computation
+    alice, bob = get_clients(2)
```
Comment (gmuraru) [inline tests/sympc/tensor/mpc_tensor_test.py:207]: Q: Why not ```parites = get_clients(2)```? (directly)

File: tests/sympc/tensor/mpc_tensor_test.py (hunk 8 above)
Comment (kamathhrishi) [reply tests/sympc/tensor/mpc_tensor_test.py:37]: Remove what exactly?

File: tests/sympc/tensor/mpc_tensor_test.py (hunk 7 above)
Comment (kamathhrishi) [reply tests/sympc/tensor/mpc_tensor_test.py:27]: Sorry for not making it clear. This was supposed to test for setup_mpc() not being called on the session.

File: tests/sympc/tensor/mpc_tensor_test.py (hunk 8 above)
Comment (gmuraru) [reply tests/sympc/tensor/mpc_tensor_test.py:27]: Q: @kamathhrishi I am not sure if you saw this, I think we can remove it

File: tests/sympc/tensor/mpc_tensor_test.py [hunk 10]
```diff
@@ -12,15 +12,28 @@
...
+        MPCTensor(secret=torch.Tensor([1, -2]), session=session)
+
+
+def test_mpc_share_nosession_exception(get_clients) -> None:
+    alice_client, bob_client = get_clients(2)
+
+    x_secret = torch.Tensor([[0.1, -1], [-4, 4]])
+    y_secret = torch.Tensor([[4.0, -2.5], [5, 2]])
+
+    with pytest.raises(ValueError):
```
Comment (gmuraru) [inline tests/sympc/tensor/mpc_tensor_test.py:32]: Q: I think you are testing the same thing?

**Review by gmuraru — CHANGES_REQUESTED** [review]
Only some questions

File: tests/sympc/tensor/share_tensor_test.py [hunk 11]
```diff
@@ -90,6 +90,28 @@ def test_reverse_ops_share_tensor_local(op_str, precision, base) -> None:
     assert np.allclose(tensor_decoded, expected_res, rtol=base ** -precision)
 
 
+def test_invalid_op_exception(op_str="truediv", precision=2, base=16) -> None:
```
Comment (gmuraru) [inline tests/sympc/tensor/share_tensor_test.py:93]: MAYBE: Remove the arguments - you are not using ```pytest.parametrize```

**Review by gmuraru — APPROVED** [review]
Great work! Thank you!

**Pull request merged**